### <b>Exercice 3.1 : MLP avec données MNIST, avec librairie Scikit-Learn</b> 

Appliquer la fonction MLPClassifier de Scikiyt-Learn aux données MNIST, en essayer d'obtenir la meilleure précision possible tout en gardant une durée d'apprentissage raisonnable.

In [16]:
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score
from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split
import numpy as np

#chargement des données
mnist = fetch_openml('mnist_784', version=1)
X, y = mnist['data'], mnist['target'].astype(np.int)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

X_train = X_train / 255.0
X_test = X_test / 255.0

#Initialisation du modèle MLP
mlp = MLPClassifier(hidden_layer_sizes=(30, 20), max_iter=20, alpha=1e-5, solver='sgd', random_state=1)

#entraînement du modèle
mlp.fit(X_train, y_train)


y_pred = mlp.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f'Précision: {accuracy * 100:.2f}%')
print(f'Durée d\'entraînement: {mlp.n_iter_} itérations')

/opt/conda/lib/python3.10/site-packages/sklearn/datasets/_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(
/tmp/ipykernel_64/4286784621.py:9: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  X, y = mnist['

Précision: 92.36%
Durée d'entraînement: 20 itérations


#### <b>Commentaires :</b>

La précision est 92.36% avec un apprentissage de 20 itérations, ce qui nous parait raisonnable.

### <b>Exercice 3.2 : MLP avec données MNIST, avec librairie Keras</b>

Essayer d'améliorer la précision du réseau de neurones (accuracy) utilisé sur les données MNIST dans le notebook principal, en limitant le nombre de paramètres du réseau de neurones. On peut jouer pour ça sur :

- la structure du réseau de neurones (nombre de couches cachées, nombre de neurones dans les couches)
- l'ajout de dropout
- la normalisation des données
- ...

In [17]:
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout
from keras.utils import np_utils

#Chargement des données
(X_train, y_train), (X_test, y_test) = mnist.load_data()
X_train = X_train.reshape(X_train.shape[0], X_train.shape[1] * X_train.shape[2])
X_test = X_test.reshape(X_test.shape[0], X_test.shape[1] * X_test.shape[2])

X_train = X_train.astype('float32') / 255
X_test = X_test.astype('float32') / 255

y_train = np_utils.to_categorical(y_train, 10)
y_test = np_utils.to_categorical(y_test, 10)

#Création du modèle
model = Sequential()
model.add(Dense(256, activation='relu', input_shape=(784,)))
model.add(Dropout(0.3))  #Ajout de Dropout pour éviter l'overfitting
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.3))
model.add(Dense(10, activation='softmax'))  # Couche de sortie avec 10 classes

#Compilation du modèle
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

# Entraînement du modèle
history = model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1, validation_data=(X_test, y_test))

# évaluation du modèle sur les données de test
score = model.evaluate(X_test, y_test, verbose=0)
print(f'Précision sur l\'ensemble de test : {score[1] * 100:.2f}%')

2024-11-16 14:39:50.657390: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-11-16 14:39:51.751546: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
2024-11-16 14:39:53.154148: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1635] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 7962 MB memory:  -> device: 0, name: NVIDIA A100 80GB PCIe MIG 1g.10gb, pci bus id: 0000:c3:00.0, compute capability: 8.0


Epoch 1/20


2024-11-16 14:39:54.818185: I tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:637] TensorFloat-32 will be used for the matrix multiplication. This will only be logged once.
2024-11-16 14:39:54.909310: I tensorflow/compiler/xla/service/service.cc:169] XLA service 0x7cc9dc048960 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
2024-11-16 14:39:54.909378: I tensorflow/compiler/xla/service/service.cc:177]   StreamExecutor device (0): NVIDIA A100 80GB PCIe MIG 1g.10gb, Compute Capability 8.0
2024-11-16 14:39:54.917095: I tensorflow/compiler/mlir/tensorflow/utils/dump_mlir_util.cc:269] disabling MLIR crash reproducer, set env var `MLIR_CRASH_REPRODUCER_DIRECTORY` to enable.
2024-11-16 14:39:54.949306: I tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:424] Loaded cuDNN version 8700
2024-11-16 14:39:55.122018: I ./tensorflow/compiler/jit/device_compiler.h:180] Compiled cluster using XLA!  This line is logged at most once for the lifetim

469/469 [==============================] - 4s 4ms/step - loss: 0.3784 - accuracy: 0.8844 - val_loss: 0.1352 - val_accuracy: 0.9568
Epoch 2/20
469/469 [==============================] - 2s 4ms/step - loss: 0.1629 - accuracy: 0.9517 - val_loss: 0.0932 - val_accuracy: 0.9710
Epoch 3/20
469/469 [==============================] - 2s 4ms/step - loss: 0.1230 - accuracy: 0.9632 - val_loss: 0.0784 - val_accuracy: 0.9750
Epoch 4/20
469/469 [==============================] - 2s 3ms/step - loss: 0.1007 - accuracy: 0.9695 - val_loss: 0.0719 - val_accuracy: 0.9780
Epoch 5/20
469/469 [==============================] - 1s 3ms/step - loss: 0.0858 - accuracy: 0.9737 - val_loss: 0.0723 - val_accuracy: 0.9778
Epoch 6/20
469/469 [==============================] - 1s 3ms/step - loss: 0.0759 - accuracy: 0.9761 - val_loss: 0.0637 - val_accuracy: 0.9803
Epoch 7/20
469/469 [==============================] - 2s 4ms/step - loss: 0.0676 - accuracy: 0.9787 - val_loss: 0.0694 - val_accuracy: 0.9776
Epoch 8/20
469/46

#### <b>Commentaires :</b>

Nous avons commencé par normaliser les données. Le réseau de neurones utilise donc 256 neurones dans la première couche et 128 neurones dans la deuxième couche. Nous avons ajouté un dropout de 30% (0.3) et nous obtenons une meilleur précision : 98.31%

### <b>Exercice 3.3 : MLP avec données MNIST, avec librairie Pytorch</b> 

Idem exercice que 3.2 mais en Pytorch.

In [18]:
import torch
import torch.nn as nn
import torch.optim as optim
from torchvision import datasets, transforms
from torch.utils.data import DataLoader

# Transformation des données
transform = transforms.Compose([
    transforms.ToTensor(), # Conversion des images en tenseurs
    transforms.Normalize((0.1307,), (0.3081,)) #normalisation pour moyenne nulle et variance de 1
])

#Chargement des données MNIST
train_dataset = datasets.MNIST(root='./data', train=True, transform=transform, download=True)
test_dataset = datasets.MNIST(root='./data', train=False, transform=transform, download=True)

#Définition des DataLoader pour l'entraînement et les tests
train_loader = DataLoader(train_dataset, batch_size=128, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=128, shuffle=False)

#Définition du modèle
class MLP(nn.Module):
    def __init__(self):
        super(MLP, self).__init__()
        self.fc1 = nn.Linear(28*28, 256) # Couche cachée avec 256 neurones
        self.dropout1 = nn.Dropout(0.3) # Dropout de 30% pour éviter l'overfitting
        self.fc2 = nn.Linear(256, 128) # Couche cachée avec 128 neurones
        self.dropout2 = nn.Dropout(0.3) # Dropout de 30% pour la deuxième couche
        self.fc3 = nn.Linear(128, 10) # Couche de sortie avec 10 classes

    def forward(self, x):
        x = x.view(-1, 28*28)

#### <b>Commentaires :</b>

Nous avons repris le modèle de l'exercice 3.2 (nombre de neurones et dropout). Nous avons normalisé les données et nous avons optimisé avec Adam.

### <b>Exercice 3.4 : MLP avec données GTSRB, avec librairie Pytorch</b> 

Appliquer aux données GTSRB le modèle étudié ci-dessus avec les données MNIST.

Par rapport à l'exemple avec les données MNIST, il faudra faire quelques adaptations : 
- les images GTSRB étant des images en couleurs, il faudra les convertir en niveaux de gris car le MLP ne gère par les canaux couleur des images
- adapter le nombre de classes
- adapter la dimension des images (on pourra prendre par exemple 48x48 ou 32x32)
- dans la fonction de chargement des images, ajouter une conversion en float (ex : data = np.array(images, dtype='float32') où images est la liste des images)
- séparer données de test et d'apprentissage par exemple à l'aide de la fonction "train_test_split" de Scikit-Learn
- changer l'optimiseur utilisé pour l'apprentissage (ex. "Adam")
- on peut également limiter le nombre d'images par classe (ex. 00), pour limiter les temps de traitement
- ...

L'objectif est d'atteindre une précision de <b>94%</b> avec un modèle possédant <b>moins de 500000 paramètres</b>.  

In [ ]:
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader, Dataset
from sklearn.model_selection import train_test_split
from torchvision import transforms
import numpy as np
import os
from PIL import Image

# Définition des classes pour GTSRB (43 classes)
NUM_CLASSES = 43

# Dimensions des images après redimensionnement
IMAGE_SIZE = 48  # Augmenté à 48x48 pour plus de précision

# Classe pour charger le dataset GTSRB
class GTSRBDataset(Dataset):
    def __init__(self, images, labels, transform=None):
        self.images = images
        self.labels = labels
        self.transform = transform

    def __len__(self):
        return len(self.images)

    def __getitem__(self, idx):
        image = self.images[idx]
        label = self.labels[idx]
        
        # Conversion en image PIL
        image = Image.fromarray(image)
        
        if self.transform:
            image = self.transform(image)
        
        return image, label

# Charger les données GTSRB
def load_gtsrb_data(data_dir, limit_per_class=None):
    images = []
    labels = []
    
    # Extensions d'image acceptées
    valid_extensions = ('.ppm', '.jpg', '.jpeg', '.png')
    
    for class_id in range(NUM_CLASSES):
        class_dir = os.path.join(data_dir, f'{class_id:05d}')
        files = os.listdir(class_dir)
        
        if limit_per_class:
            files = files[:limit_per_class]
        
        for file_name in files:
            if file_name.lower().endswith(valid_extensions):
                file_path = os.path.join(class_dir, file_name)
                try:
                    image = Image.open(file_path).convert('L')  # Convertir en niveaux de gris
                    image = np.array(image.resize((IMAGE_SIZE, IMAGE_SIZE)), dtype='float32')
                    images.append(image)
                    labels.append(class_id)
                except Exception as e:
                    print(f"Fichier ignoré (non valide): {file_path}, Erreur: {e}")
    
    return np.array(images), np.array(labels)

# Charger les données GTSRB
data_dir = '/home/jovyan/iadatasets/GTSRB/Final_Training/Images/'
images, labels = load_gtsrb_data(data_dir, limit_per_class=500)    

# Normalisation des données
images = images / 255.0  # Normalisation entre 0 et 1

# Séparer les données en ensembles d'entraînement et de test
X_train, X_test, y_train, y_test = train_test_split(images, labels, test_size=0.2, random_state=42)

# Transformation des images
transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize((0.5,), (0.5,))  # Normalisation autour de 0
])

# Création des datasets et loaders
train_dataset = GTSRBDataset(X_train, y_train, transform=transform)
test_dataset = GTSRBDataset(X_test, y_test, transform=transform)

train_loader = DataLoader(train_dataset, batch_size=128, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=128, shuffle=False)

# Définition du modèle MLP pour GTSRB
class MLP(nn.Module):
    def __init__(self):
        super(MLP, self).__init__()
        self.fc1 = nn.Linear(IMAGE_SIZE * IMAGE_SIZE, 1024)  # Augmentation à 1024 neurones
        self.dropout1 = nn.Dropout(0.3)
        self.fc2 = nn.Linear(1024, 512)  # Deuxième couche avec 512 neurones
        self.dropout2 = nn.Dropout(0.3)
        self.fc3 = nn.Linear(512, 256)  # Troisième couche avec 256 neurones
        self.dropout3 = nn.Dropout(0.3)
        self.fc4 = nn.Linear(256, NUM_CLASSES)  # Couche de sortie

    def forward(self, x):
        x = x.view(-1, IMAGE_SIZE * IMAGE_SIZE)  # Aplatir les images
        x = torch.relu(self.fc1(x))
        x = self.dropout1(x)
        x = torch.relu(self.fc2(x))
        x = self.dropout2(x)
        x = torch.relu(self.fc3(x))
        x = self.dropout3(x)
        x = self.fc4(x)
        return torch.log_softmax(x, dim=1)

# Initialisation du modèle, de la fonction de coût et de l'optimiseur
model = MLP()
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

# Entraînement du modèle
def train(model, device, train_loader, optimizer, criterion, epoch):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = criterion(output, target)
        loss.backward()
        optimizer.step()
        if batch_idx % 100 == 0:
            print(f'Epoch {epoch} [{batch_idx * len(data)}/{len(train_loader.dataset)}]\tLoss: {loss.item():.6f}')        
            
# Test du modèle
def test(model, device, test_loader, criterion):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += criterion(output, target).item()
            pred = output.argmax(dim=1, keepdim=True)
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)
    accuracy = 100. * correct / len(test_loader.dataset)
    print(f'\nTest set: Average loss: {test_loss:.4f}, Accuracy: {correct}/{len(test_loader.dataset)} ({accuracy:.2f}%)\n')

# Entraînement et test du modèle
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model.to(device)
num_epochs = 97
for epoch in range(1, num_epochs + 1):
    train(model, device, train_loader, optimizer, criterion, epoch)
    test(model, device, test_loader, criterion)

Epoch 1 [0/14366]	Loss: 3.761223
Epoch 1 [12800/14366]	Loss: 2.294377

Test set: Average loss: 0.0171, Accuracy: 1420/3592 (39.53%)

Epoch 2 [0/14366]	Loss: 2.108185
Epoch 2 [12800/14366]	Loss: 1.590467

Test set: Average loss: 0.0102, Accuracy: 2140/3592 (59.58%)

Epoch 3 [0/14366]	Loss: 1.459478
Epoch 3 [12800/14366]	Loss: 1.193142

Test set: Average loss: 0.0084, Accuracy: 2498/3592 (69.54%)

Epoch 4 [0/14366]	Loss: 1.259585
Epoch 4 [12800/14366]	Loss: 1.079097

Test set: Average loss: 0.0061, Accuracy: 2817/3592 (78.42%)

Epoch 5 [0/14366]	Loss: 0.732259


#### <b>Commentaires :</b>

Notre code utilise un MLP avec PyTorch pour classifier les images du dataset GTSRB après les avoir converties en niveaux de gris et redimensionnées. Cependant, nous n'avons pas atteint l'objectif de 94% de précision avec moins de 500 000 paramètres (nous avons environ 93% avec 100 epochs).